java/semantickernel-api/src/main/java/com/microsoft/semantickernel/FunctionInvocation.java

```diff
@@ -0,0 +1,191 @@
+import com.microsoft.semantickernel.orchestration.ToolCallBehavior.UnmodifiableToolCallBehavior;
+import com.microsoft.semantickernel.orchestration.contextvariables.ContextVariable;
+import com.microsoft.semantickernel.orchestration.contextvariables.ContextVariableType;
+import java.util.function.BiConsumer;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.CoreSubscriber;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.SynchronousSink;
+
+public class FunctionInvocation<T> extends Mono<FunctionResult<T>> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FunctionInvocation.class);
+
+    private final KernelFunction<?> function;
+    private final Kernel kernel;
+    @Nullable
+    private KernelFunctionArguments arguments;
+    @Nullable
+    private ContextVariableType<T> resultType;
+    @Nullable
+    private UnmodifiableKernelHooks hooks;
+    @Nullable
+    private PromptExecutionSettings promptExecutionSettings;
+    @Nullable
+    private UnmodifiableToolCallBehavior toolCallBehavior;
+
+
+    public FunctionInvocation(
+        Kernel kernel,
+        KernelFunction<T> function) {
+        this.function = function;
+        this.kernel = kernel;
+    }
+
+    public FunctionInvocation(
+        Kernel kernel,
+        KernelFunction<?> function,
+        @Nullable
+        ContextVariableType<T> resultType) {
+        this.function = function;
+        this.kernel = kernel;
+        this.resultType = resultType;
+    }
+
+    public FunctionInvocation<T> withArguments(
+        @Nullable
+        KernelFunctionArguments arguments) {
+        if (arguments == null) {
+            this.arguments = null;
+        } else {
+            this.arguments = new KernelFunctionArguments(arguments);
+        }
+        return this;
+    }
+
+    public <U> FunctionInvocation<U> withResultType(ContextVariableType<U> resultType) {
+        return new FunctionInvocation<>(
+            kernel,
+            function,
+            resultType)
+            .withArguments(arguments)
+            .withKernelHooks(hooks)
+            .withPromptExecutionSettings(promptExecutionSettings)
+            .withToolCallBehavior(toolCallBehavior);
+    }
+
+    public FunctionInvocation<T> withKernelHooks(
+        @Nullable
+        KernelHooks hooks) {
+        this.hooks = unmodifiableClone(hooks);
+        return this;
+    }
+
+    public FunctionInvocation<T> withPromptExecutionSettings(
+        @Nullable
+        PromptExecutionSettings promptExecutionSettings) {
+        this.promptExecutionSettings = promptExecutionSettings;
+        return this;
+    }
+
+    public FunctionInvocation<T> withToolCallBehavior(@Nullable ToolCallBehavior toolCallBehavior) {
+        this.toolCallBehavior = unmodifiableClone(toolCallBehavior);
+        return this;
+    }
+
+    @Nullable
+    private static UnmodifiableToolCallBehavior unmodifiableClone(
```

Comment:
Just a nit... can the private methods come after the public and protected methods?